### PR TITLE
filter duplicates on g:ctrlp_lines while preserving order

### PR DESCRIPTION
**Pull request by timfeirg**
This closes https://github.com/ctrlpvim/ctrlp.vim/issues/314, https://github.com/ctrlpvim/ctrlp.vim/issues/168

This may not be the best implementation, but it should only affect performance when refreshing a very large `:CtrlPMixed` lists.

I have very little experience with vimL, the project code hasn't been easy for me to understand, I'm not very sure where and how to implement this at all, I'll be glad to take any advises and fix up my commit.

@mattn

### CODE REVIEW

**Comment by timfeirg** [comment]
request for comment @tacahiroy @mattn

**Comment by timfeirg** [comment]
ping @ludovicchabant

**Comment by tacahiroy** [comment]
hmm - i still don't reproduce the issue actually

**Comment by timfeirg** [comment]
This would appear when one file appears in both files and MRU files at the same time.

steps to reproduct:

* navigate to a brand new directory with some files, not cached by ctrlp
* open up vim with ctrlp loaded
* `:CtrlpMixed` and open up one of the files
* re-open vim and launch `:CtrlpMixed` again,  the file from the previous step will be duplicated

image

@tacahiroy

**Comment by mattn** [comment]
Current implementation doesn't remove duplicates.

**Comment by timfeirg** [comment]
I've been using my own fork without a problem, didn't notice any performance regression, would you be interested in merging this PR? I mean, until a better implementation is introduced? @mattn 

I think it's probably best to filter out duplicates just before the ctrlp pane is rendered, but I haven't been able to pinpoint the code that does this.

**Comment by mattn** [comment]
I want to look your diff. And figure out what is problem.

**Comment by timfeirg** [comment]
Dear developers, since https://github.com/ctrlpvim/ctrlp.vim/pull/370 couldn't resolve this issue, would you be interested in merging this PR for the time being? We can, of course, revert this PR when a better implementation can be introduced.

Without duplicate filtering, `:CtrlPMixed` could display quite a lot of useless information, considering the default ctrlp pane is only 10 lines long, this behavior is very much wanted.

@mattn

**Comment by mattn** [comment]
Before merging, I want to reproduce this.

**Comment by timfeirg** [comment]
I'm on e66b5b50e5c539354813e2ecac6fb75a2121e29e and I can still reproduce this following the steps in https://github.com/ctrlpvim/ctrlp.vim/pull/367#issuecomment-303430768

File: autoload/ctrlp/mixed.vim [hunk 1]
```diff
@@ -64,6 +64,8 @@ fu! s:getnewmix(cwd, clim)
 	if len(g:ctrlp_lines) <= a:clim
 		cal sort(g:ctrlp_lines, 'ctrlp#complen')
 	en
+	" remove duplicates from the final result set
+  let g:ctrlp_lines=filter(copy(g:ctrlp_lines), 'index(g:ctrlp_lines, v:val, v:key+1)==-1')
```
Comment (mattn) [inline autoload/ctrlp/mixed.vim]: indentation seems to be broken.

**Pull request closed without merging**